stream/statelib/src/main/java/org/apache/bookkeeper/statelib/impl/rocksdb/checkpoint/RocksdbRestoreTask.java

```diff
@@ -67,65 +67,10 @@ public void restore(String checkpointId, CheckpointMetadata metadata) throws Sta
         }
     }
 
-    private List<String> getFilesToCopy(String checkpointId,
-                                        File checkpointedDir,
-                                        CheckpointMetadata metadata) throws IOException {
-        if (!checkpointedDir.exists()) {
-            Files.createDirectories(
-                Paths.get(checkpointedDir.getAbsolutePath()));
-        }
-
-        List<String> filesToCopy = Lists.newArrayListWithExpectedSize(metadata.getFilesCount());
-        for (String fileName : metadata.getFilesList()) {
-            File localFile = new File(checkpointedDir, fileName);
-            if (!localFile.exists()) {
-                filesToCopy.add(fileName);
-                continue;
-            }
-
-            String srcFile;
-            if (RocksUtils.isSstFile(localFile)) {
-                srcFile = RocksUtils.getDestSstPath(dbPrefix, localFile);
-            } else {
-                srcFile = RocksUtils.getDestPath(dbPrefix, checkpointId, localFile);
-            }
-
-            long srcFileLength = checkpointStore.getFileLength(srcFile);
-            long localFileLength = localFile.length();
-            if (srcFileLength != localFileLength) {
-                filesToCopy.add(fileName);
-            }
-        }
-
-        return filesToCopy;
-    }
-
     private void copyFilesFromRemote(String checkpointId,
-                                     File checkpointedDir,
-                                     List<String> remoteFiles) throws IOException {
-        for (String file : remoteFiles) {
-            copyFileFromRemote(checkpointId, checkpointedDir, file);
+                                     List<CheckpointFile> remoteFiles) throws IOException {
+        for (CheckpointFile file : remoteFiles) {
+            file.copyFromRemote(checkpointStore, dbPrefix, checkpointId);
```

Comment:
I am probably missing something obvious.
I expected checksum validation after file copy from remote, is it happening at some other place (implicitly via equals, as part of some map/set check?)?

Reply:
The rocksdb already has validation to verify the contents of the SST files. Here we want to ensure that we use the correct SST file with the checkpoint. Essentially we are giving the SST files a unique name based on their contents.

Reply:
I see, thank you. 

Let's say the name+checksum gets added to both FileInfos and Files.
Will this be good enough for the older bookie to use the checkpoint?

Reply:
No, that will not work. The files need to get renamed to *.sst for the rocksdb restore process to work. Also there are other places where we check if the name ends with ".sst" to determine if this is a `sst` file.

Reply:
Let me see if I can add some clarity.
* Restore code get the list of files from the `metadata` file (which is the first file we restore). All files other than SST files are stored under the checkpoint directory. SST files however are stored under common "ssts" directory. The code uses the extension of the file to determine if the the file is SST or nor. If it is a SST file, it will copy it from common "ssts" folder. If we add checksum to names of file added to `files` field in metadata, this code will fail and will not be able to copy the files from the checkpoint store.
* Now we can work around the above by prefixing the checksum (e.g 00009.sst -> {checksum}_00009.sst). The code will correctly detect this as a SST file and copy it from the common "ssts" folder. These files however would still need to be renamed ({checksum}_00009.sst -> 00009.sst) locally. This is the files that rocksdb will be looking for. Since there is no code in the previous version that can do that, it will not work.

Reply:
would it help if we used relative path there? like "./{checksum}/../00009.sst"?
I am not sure if [normalize()](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#normalize--) happens automatically or not when the new file created.

Reply:
I don't think that will help here.